_plugins/events.rb

```diff
@@ -0,0 +1,124 @@
+require 'date'
+require 'icalendar'
+
+module Jekyll
+  module Events
+
+    # Helpers
+
+    def date_or_datetime(d)
+      # Input could be a String, Date, or DateTime
+      # Want to return a Date or DateTime since they have mostly the same methods
+      d.kind_of?(Date) ? d : d.to_datetime
+    rescue
+      nil
+    end
+
+    def ical_time(d)
+      # Dates and DateTimes have different handling so can't use date_or_datetime
+      if d.class == Date # no time
+        Icalendar::Values::Date.new(d.strftime('%Y%m%d'))
+      else # with time, String or DateTime object
+        Icalendar::Values::DateTime.new(d.to_datetime.utc.strftime('%Y%m%dT%H%M%SZ'))
+      end
+    end
+
+    # Filters
+
+    def filter_old_events(events)
+      events.reject { |event| date_or_datetime(event['start']).past? }
+    end
+
+    def filter_json_events(events)
+      events.map do |event|
+        # for calendar rendering, link within the page for current events
+        event['url'] = "/events/##{Jekyll::Utils.slugify(event['title'])}" unless
+          date_or_datetime(event['start']).past?
+        event
+      end.to_json
+    end
+
+    def fix_calendar_dates(events)
+      events.map do |event|
+        dstart = date_or_datetime(event['start'])
+        dend   = date_or_datetime(event['end'])
+
+        # Calendar assumes midnight is at the *start* of the day, so a date-only
+        # range like 01-30 -> 01-31 would only render on the 30th.
+        # So add a day for date-only events
+        if (event['start'].class == Date) && !dend.nil?
+          dend   = dend.next_day
+        end
+
+        # Calendar requires JSON like '2016-04-06T15:45:00+01:00' or '2016-04-06'
+        event['start'] = dstart.to_s
+        event['end']   = dend.to_s unless dend.nil?
+        event
+      end
+    end
+
+    def format_event_date start_date, end_date
+      # Compares start_date with end_date to decide how to format the date correctly
+      return "" unless start_date
+
+      if start_date.class == Date
+        # no times, just use dates
+        if end_date.nil? || start_date == end_date
+          # one-day event
+          return start_date.strftime('%-d %B, %Y')
```

Comment:
Useless return? There are retuns all over the place in here which I don't think are needed - also you may split this in multiple functions or a date/time parser class